### ISSUE TITLE
Two invoices to the entire order

### ISSUE THREAD

**Issue by kassner** [magento/magento2#1464]
Trying a simple snippet:

```php
$this->_objectManager->create('Magento\Sales\Model\Order')->load(1);

$order->prepareInvoice()->save();
$order->save();

$order->prepareInvoice()->save();
$order->save();
```

In such case, `prepareInvoice` generates an invoice with all items from the order. In the second run, the invoice is generated again with all items (even those that already were invoiced). The shipping amount is invoiced only in the first time.

Is this an expected behavior? I guess it should not allow to invoice more items than were ordered.

**Comment by vpelipenko** [comment]
We can call it a bug. But could you describe here business case? When can we face with such behavior?
Internal ticket: MAGETWO-39989

**Comment by kassner** [comment]
Sorry, I was calling it in the wrong way, need to `register` it before saving.

Code for future reference:

```php
    protected function invoice(OrderInterface $order)
    {
        if (!$order->canInvoice()) {
            throw new \Magento\Framework\Exception\LocalizedException(
                __('The order does not allow an invoice to be created.')
            );
        }

        /** @var \Magento\Sales\Model\Order\Invoice $invoice */
        $invoice = $this->_objectManager->create('Magento\Sales\Model\Service\Order', ['order' => $order])->prepareInvoice();

        if (!$invoice) {
            throw new \Magento\Framework\Exception\LocalizedException(__('We can\'t save the invoice right now.'));
        }

        if (!$invoice->getTotalQty()) {
            throw new \Magento\Framework\Exception\LocalizedException(
                __('You can\'t create an invoice without products.')
            );
        }

        $invoice->register();

        $invoice->getOrder()->setCustomerNoteNotify(true);
        $invoice->getOrder()->setIsInProcess(true);

        $transactionSave = $this->_objectManager->create('Magento\Framework\DB\Transaction')
            ->addObject($invoice)
            ->addObject($invoice->getOrder());

        $transactionSave->save();
    }
```

Thanks!

**Comment by magento-engcom-team** [comment]
Hi @kassner. Thank you for your report. 
The issue has been fixed in magento/magento2#29058 by @gabrieldagama in 2.4-develop branch
Related commit(s):
 - [dfc733a98eb96cf685938a9cfdb1e10c3aeb1ee1](https://github.com/magento/magento2/commit/dfc733a98eb96cf685938a9cfdb1e10c3aeb1ee1)
 - [3a1a3c73bdcf29835641572bc4e51ece2941db33](https://github.com/magento/magento2/commit/3a1a3c73bdcf29835641572bc4e51ece2941db33)
 - [15aa96f9ee6bebebe8d3102c39e655553ef21ee0](https://github.com/magento/magento2/commit/15aa96f9ee6bebebe8d3102c39e655553ef21ee0)
 - [1fcb11a63b832c4811621c636e2571ed261e27ac](https://github.com/magento/magento2/commit/1fcb11a63b832c4811621c636e2571ed261e27ac)
 - [a7e980c8c37b7caebf8500f906b010ea64abc832](https://github.com/magento/magento2/commit/a7e980c8c37b7caebf8500f906b010ea64abc832)
 - [279cf2df2432dc71a6fdab767bcefd8dd91909c8](https://github.com/magento/magento2/commit/279cf2df2432dc71a6fdab767bcefd8dd91909c8)
 - [11082e79ed5547da5bd9db408c987ae1e3374722](https://github.com/magento/magento2/commit/11082e79ed5547da5bd9db408c987ae1e3374722)
 - [e0a777b1adaff963434d7acc228b897d16ddb762](https://github.com/magento/magento2/commit/e0a777b1adaff963434d7acc228b897d16ddb762)
 - [89551473a49496ac547d27bf8ffc69981f976389](https://github.com/magento/magento2/commit/89551473a49496ac547d27bf8ffc69981f976389)
 - [545f886e6b97b30f083592d05c589239aa07ab6b](https://github.com/magento/magento2/commit/545f886e6b97b30f083592d05c589239aa07ab6b)
 - [5f82525415fb2cc76ee4dc08e9cc0ce995762ffd](https://github.com/magento/magento2/commit/5f82525415fb2cc76ee4dc08e9cc0ce995762ffd)
 - [aa11cf598727bffcc356ef44e11d1825d2075569](https://github.com/magento/magento2/commit/aa11cf598727bffcc356ef44e11d1825d2075569)
 - [752eb5af9e1581c8a1508d63390d1fb07ea3ddae](https://github.com/magento/magento2/commit/752eb5af9e1581c8a1508d63390d1fb07ea3ddae)
 - [9dffdc53f0e415e4f664b26861148136218b359c](https://github.com/magento/magento2/commit/9dffdc53f0e415e4f664b26861148136218b359c)
 - [f02a540efc7863011743f1878cd7b1c6b668aee6](https://github.com/magento/magento2/commit/f02a540efc7863011743f1878cd7b1c6b668aee6)
 - [500acc33cdaca4ac3629e298ee23c73880f74eec](https://github.com/magento/magento2/commit/500acc33cdaca4ac3629e298ee23c73880f74eec)
 - [acf425b4171db3d9b153b153eb7e17f35af719ce](https://github.com/magento/magento2/commit/acf425b4171db3d9b153b153eb7e17f35af719ce)
 - [377ed536493b73bbdb0c58f2b6ef6723327230bf](https://github.com/magento/magento2/commit/377ed536493b73bbdb0c58f2b6ef6723327230bf)
 - [193765d016180d1cf6bcadf904d741f483f0f1ba](https://github.com/magento/magento2/commit/193765d016180d1cf6bcadf904d741f483f0f1ba)
 - [095dcc631d8e84ece779f7d4490e4a980fb61360](https://github.com/magento/magento2/commit/095dcc631d8e84ece779f7d4490e4a980fb61360)

The fix will be available with the upcoming 2.4.1 release.
